oci/layer/generate.go

```diff
@@ -97,3 +101,44 @@ func GenerateLayer(path string, deltas []mtree.InodeDelta, opt *MapOptions) (io.
 
 	return reader, nil
 }
+
+// GenerateInsertLayer generates a completely new layer from the root path to
+// be inserted into the image at "prefix".
+func GenerateInsertLayer(root string, prefix string, opt *MapOptions) io.ReadCloser {
+	root = path.Clean(root)
+	rootPrefixLen := len(root) - len(path.Base(root))
+	prefix = strings.TrimPrefix(prefix, "/")
```

Comment:
This should be some mix of `utils.CleanPath` similar. I'm not sure which. I get that you want to strip leading '/' but you can do that in the `Join` call. In addition, I'm not really sure what you're doing with `rootPrefixLen` -- why not use `path.Rel`?

Reply:
rootPrefixLen is basically some fancy math so that if the user does: `umoci insert --image ... /thing/to/insert /usr/bin`, we end up with `insert` in `/usr/bin`, and not `/usr/bin/thing/to/insert`; `path.Rel` doesn't quite do the right thing here.

The TrimPrefix bit is because of this hunk: https://github.com/openSUSE/umoci/blob/master/oci/layer/tar_generate.go#L94 which causes a failure if there's an absolute path passed to the tar generator. It's not really clear why that's there, but I assume you had a reason, so I just trimmed the prefix. I'm not sure any path cleaning will do that, though.

Reply:
> umoci insert --image ... /thing/to/insert /usr/bin, we end up with insert in /usr/bin, and not /usr/bin/thing/to/insert; path.Rel doesn't quite do the right thing here.

Shouldn't be able to just do `filepath.Base`?

> The TrimPrefix bit is because of this hunk: https://github.com/openSUSE/umoci/blob/master/oci/layer/tar_generate.go#L94 which causes a failure if there's an absolute path passed to the tar generator. It's not really clear why that's there, but I assume you had a reason, so I just trimmed the prefix.

It's for security reasons, to double-check that the path lexically doesn't have any un-clean components. But you're right that I didn't consider the `"/"+"/abc" != filepath.Join("/", "/abc")` edge-case (because generally tar archives don't have leading `/` for their entries).

We could fix this by removing leading `/` when normalising the paths.

Reply:
no, because `filepath.Base("/thing/to/insert") == "insert"`, not `"/thing/to"`, which is what we want to know the length of. I think `filepath.Dir()` is the right thing though, I can change it to that.

R.e. the tar issue, you do a CleanPath() before that check, so shouldn't all the ..s be cleaned out? I suppose extra paranoia is ok, but if you don't trust your own code...

I'm happy to just trim the leading / in normalise(), though.